### PR TITLE
ERF: V2.2 file support

### PR DESCRIPTION
**Pull request by Nostritius**
This PR adds support for V2.2 ERF files and their compression.

### CODE REVIEW

File: src/common/deflate.cpp [hunk 1]
```diff
@@ -60,6 +60,28 @@ static void initZStream(z_stream &strm, int windowBits, size_t size, const byte
 		throw Exception("Could not initialize zlib inflate: %s (%d)", zError(zResult), zResult);
 }
 
+static void initDeflateZStream(z_stream &strm, int windowBits, size_t size, const byte *data) {
+	/* Initialize the zlib data stream for decompression with our input data. */
```
Comment (DrMcCoy) [inline src/common/deflate.cpp]: compression, not decompression :P

File: src/common/deflate.h [hunk 2]
```diff
@@ -117,6 +117,33 @@ SeekableReadStream *decompressDeflateWithoutOutputSize(ReadStream &input, size_t
 size_t decompressDeflateChunk(SeekableReadStream &input, int windowBits, byte *output, size_t outputSize,
                               unsigned int frameSize = 4096);
 
+/** Compress (deflate) using zlib's DEFLATE algorithm.
+ *
```
Comment (DrMcCoy) [inline src/common/deflate.h]: Trailing whitespace

File: src/aurora/erfwriter.cpp [hunk 3]
```diff
@@ -113,12 +116,63 @@ ERFWriter::ERFWriter(uint32 id, uint32 fileCount, Common::SeekableWriteStream &s
...
+			// Write unknown 0xFFFFFFFF value
+			stream.writeUint32BE(0xFFFFFFFF);
+
+			// Create and write the flags.
+			uint32 flags = 0;
+
+			uint32 compressionFlags = 0;
+			switch (compression) {
+				case kCompressionBiowareZlib:
+					compressionFlags |= 0b00100000000000000000000000000000;
```
Comment (DrMcCoy) [inline src/aurora/erfwriter.cpp]: Binary literals are a C++14 feature, while xoreos-tools is still C++11

File: man/erf.1 [hunk 4]
```diff
@@ -24,15 +24,17 @@ This tool supports version V1.0, as used by
 and
 .Em The Witcher
 and version V2.0 as used by
-.Em Dragon Age: Origins .
+.Em Dragon Age: Origins
+and version V2.2 as used by
+.Em Dragon Age: Origins
+and
+.Em Dragon Age 2
```
Comment (DrMcCoy) [inline man/erf.1]: Since you're changing the man page, please also update its date at the top of the file

File: src/erf.cpp [hunk 5]
```diff
@@ -127,6 +133,16 @@ bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue
...
+	parser.addOption("v22", "Generate a V2.2 ERF file",
+					 kContinueParsing,
+					 makeAssigners(new ValAssigner<Aurora::ERFWriter::Version>(Aurora::ERFWriter::kERFVersion22, version)));
+	parser.addSpace();
+	parser.addOption("bzlib", "Compress using BioWare zlib method",
+					 kContinueParsing,
+					 makeAssigners(new ValAssigner<Aurora::ERFWriter::Compression>(Aurora::ERFWriter::kCompressionBiowareZlib, compression)));
+	parser.addOption("zlib", "Compress using headerless zlib method",
+					 kContinueParsing,
+					 makeAssigners(new ValAssigner<Aurora::ERFWriter::Compression>(Aurora::ERFWriter::kCompressionHeaderlessZlib, compression)));
```
Comment (DrMcCoy) [inline src/erf.cpp]: Please uses spaces for alignment. See lines 134 and 135 above

File: src/erf.cpp [hunk 6]
```diff
@@ -58,11 +59,15 @@ int main(int argc, char **argv) {
 		uint32 id = kERFID;
 		Common::UString archive;
 		Aurora::ERFWriter::Version version = Aurora::ERFWriter::kERFVersion10;
+		Aurora::ERFWriter::Compression compression = Aurora::ERFWriter::kCompressionNone;
 		std::set<Common::UString> files;
 
-		if (!parseCommandLine(args, returnValue, archive, files, version, id, game))
+		if (!parseCommandLine(args, returnValue, archive, files, version, compression, id, game))
```
Comment (DrMcCoy) [inline src/erf.cpp]: You're not passing that compression value to Aurora::ERFWriter in line 74, though

File: src/aurora/erfwriter.cpp [hunk 7]
```diff
@@ -113,12 +116,63 @@ ERFWriter::ERFWriter(uint32 id, uint32 fileCount, Common::SeekableWriteStream &s
 			break;
 		}
 
+		case kERFVersion22: {
```
Comment (DrMcCoy) [inline src/aurora/erfwriter.cpp]: The constructor and also the method `add()` are getting a bit long. Maybe it's time to split the version bits into individual private functions, like the ERF reader does? I.e. `initV10()`/`initV20()`/`initV22()` and `addV10()`/`addV20()`/`addV22()`

**Comment by Nostritius** [comment]
The issues should now be fixed

**Comment by DrMcCoy** [comment]
Merged as 3067c50...4e6dee8, thanks! :)

**Pull request closed without merging**